bigquery_etl/shredder/config.py

```diff
@@ -474,3 +475,85 @@ def find_experiment_analysis_targets(pool, client, project=EXPERIMENT_ANALYSIS):
         client_id_target(table=qualified_table_id(table)): DESKTOP_SRC
         for table in tables
     }
+
+
+PIONEER_PROD = "moz-fx-data-pioneer-prod"
+
+
+def find_pioneer_targets(pool, client, project=PIONEER_PROD, study_projects={}):
+    """Return a dict like DELETE_TARGETS for Pioneer tables."""
+
+    def __get_tables_with_pioneer_id__(dataset_list):
+        return [
+            table
+            for table in pool.map(
+                client.get_table,
+                [
+                    table
+                    for tables in pool.map(
+                        client.list_tables,
+                        dataset_list,
+                        chunksize=1,
+                    )
+                    for table in tables
+                ],
+                chunksize=1,
+            )
+            if any(field.name == PIONEER_ID for field in table.schema)
+            and table.table_type != "VIEW"
+        ]
+
+    datasets = {
+        dataset.reference
+        for dataset in client.list_datasets(project)
+        if dataset.reference.dataset_id.startswith("pioneer_")
+    }
+    # There should be a single stable and derived dataset per study
+    stable_datasets = {dr for dr in datasets if dr.dataset_id.endswith("_stable")}
+    derived_datasets = {dr for dr in datasets if dr.dataset_id.endswith("_derived")}
+
+    stable_tables = [
+        table
+        for tables in pool.map(client.list_tables, stable_datasets, chunksize=1)
+        for table in tables
+    ]
+
+    sources = {
+        table.dataset_id: DeleteSource(qualified_table_id(table), PIONEER_ID, project)
+        # dict comprehension will only keep the last value for a given key, so
+        # sort by table_id to use the latest version
+        for table in sorted(stable_tables, key=lambda t: t.table_id)
+        if table.table_id.startswith("deletion_request_")
+    }
+
+    return {
+        **{
+            # stable tables
+            pioneer_target(
+                table=qualified_table_id(table), project=PIONEER_PROD
+            ): sources[table.dataset_id]
+            for table in stable_tables
+            if not table.table_id.startswith("deletion_request_")
+            and not table.table_id.startswith("pioneer_enrollment_")
+        },
+        **{
+            # derived tables with pioneer_id
+            pioneer_target(
+                table=qualified_table_id(table), project=PIONEER_PROD
+            ): sources[table.dataset_id]
+            for table in __get_tables_with_pioneer_id__(derived_datasets)
+        },
+        **{
+            # tables with pioneer_id located in study analysis projects
+            pioneer_target(
+                table=qualified_table_id(table), project=analysis_project
+            ): sources[study_name + "_stable"]
```

Comment:
What's the `_stable` here for? The analysis projects will only have an `analysis` dataset that needs to be inspected.

Reply:
This is used to find a matching deletion request table. I added a comment explaining this.